src/Components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,102 @@
+import React, { useState, useRef, useEffect } from 'react';
+import cn from 'classnames';
+import { Todo } from '../../types/Todo';
+import { useTodosDispatch } from '../TodoContext/TodoContext';
+import { KeyEvent } from '../../types/KeyEvent';
+import { State } from '../../types/State';
+
+type Props = {
+  todo: Todo;
+};
+
+export const TodoItem: React.FC<Props> = ({ todo }) => {
+  const editInputRef = useRef<HTMLInputElement | null>(null);
+  const dispatch = useTodosDispatch();
+  const [isEditing, setIsEditing] = useState(false);
+  const [editedTitle, setEditedTitle] = useState(todo.title);
+
+  const handleDeleteButton = () => {
+    dispatch({
+      type: State.DELETED,
+      id: todo.id,
+    });
+  };
+
+  useEffect(() => {
+    if (isEditing && editInputRef.current) {
+      editInputRef.current.focus();
+    }
+  }, [isEditing]);
+
+  const handleDoubleClick = () => {
+    setIsEditing(true);
+  };
+
+  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    switch (e.key) {
+      case KeyEvent.ENTER:
+        if (editedTitle.trim() !== '') {
+          dispatch({
+            type: State.EDIT,
+            task: { ...todo, title: editedTitle },
+          });
+        }
+
+        setIsEditing(false);
+        break;
+      case KeyEvent.ESCAPE:
+        setEditedTitle(todo.title);
+        setIsEditing(false);
+        break;
+      default:
+    }
+  };
+
+  const handleEdit = () => {
+    if (editedTitle.trim()) {
+      setEditedTitle(todo.title);
+    }
+
+    dispatch({
+      type: State.EDIT,
+      task: { ...todo, title: editedTitle },
+    });
+
+    setIsEditing(false);
+  };
+
+  return (
+    <li className={cn({ completed: todo.completed, editing: isEditing })}>
+      <div className="view" onDoubleClick={handleDoubleClick}>
+        <input
+          type="checkbox"
+          checked={todo.completed}
+          className="toggle"
+          onChange={() => {
+            dispatch({
+              type: State.EDIT,
+              task: { ...todo, completed: !todo.completed },
+            });
+          }}
```

Comment:
Move to separate function  (transfer the function to a variable, and here only write the name of the function)